File: test/easysdk-release/flutter/release_smoke_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:wukong_easy_sdk/wukong_easy_sdk.dart';

void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('released pub package completes bidirectional messaging', (
    tester,
  ) async {
    const aliceUid = String.fromEnvironment('ALICE_UID');
    const aliceToken = String.fromEnvironment('ALICE_TOKEN');
    const bobUid = String.fromEnvironment('BOB_UID');
    const aliceToBobText = String.fromEnvironment('ALICE_TO_BOB_TEXT');
    const bobToAliceText = String.fromEnvironment('BOB_TO_ALICE_TEXT');

    expect(aliceUid, isNotEmpty);
    expect(aliceToken, isNotEmpty);
    expect(bobUid, isNotEmpty);

    final sdk = WuKongEasySDK.getInstance();
    final receivedReply = Completer<Message>();
    late final WuKongEventListener<Message> listener;
    listener = (message) {
      final payload = message.payload;
      if (message.fromUid == bobUid &&
          payload is Map &&
          payload['content'] == bobToAliceText &&
          !receivedReply.isCompleted) {
        receivedReply.complete(message);
      }
    };

    sdk.addEventListener(WuKongEvent.message, listener);
    try {
      await sdk.init(
        const WuKongConfig(
          serverUrl: 'ws://127.0.0.1:5200',
          uid: aliceUid,
          token: aliceToken,
          deviceFlag: WuKongDeviceFlag.app,
        ),
      );
      await sdk.connect().timeout(const Duration(seconds: 30));
      expect(sdk.isConnected, isTrue);

      final acknowledgment = await sdk.send(
        channelId: bobUid,
        channelType: WuKongChannelType.person,
        payload: const {'type': 1, 'content': aliceToBobText},
      );
      expect(acknowledgment.messageSeq, greaterThan(0));

      final reply = await receivedReply.future.timeout(
        const Duration(seconds: 30),
      );
      expect(reply.messageSeq, greaterThan(0));
    } finally {
      sdk.removeEventListener(WuKongEvent.message, listener);
      sdk.disconnect();
    }

    await tester.pump(const Duration(milliseconds: 200));
    expect(sdk.isConnected, isFalse);
    // ignore: avoid_print
    print(
      'FLUTTER_RELEASE_SMOKE_PASS package=wukong_easy_sdk@1.1.0 '
      'alice-to-bob=true bob-to-alice=true disconnected=true',
    );
  });
}

File: test/easysdk-release/android/ReleaseSmokeTest.kt
package com.githubim.easysdk.example

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.githubim.easysdk.WuKongConfig
import com.githubim.easysdk.WuKongEasySDK
import com.githubim.easysdk.enums.WuKongChannelType
import com.githubim.easysdk.enums.WuKongDeviceFlag
import com.githubim.easysdk.enums.WuKongEvent
import com.githubim.easysdk.listener.WuKongEventListener
import com.githubim.easysdk.model.Message
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ReleaseSmokeTest {
    @Test
    fun releasedMavenPackageCompletesBidirectionalMessaging() = runBlocking {
        val arguments = InstrumentationRegistry.getArguments()
        val aliceUid = requireArgument(arguments.getString("aliceUid"), "aliceUid")
        val aliceToken = requireArgument(arguments.getString("aliceToken"), "aliceToken")
        val bobUid = requireArgument(arguments.getString("bobUid"), "bobUid")
        val aliceToBobText = requireArgument(
            arguments.getString("aliceToBobText"),
            "aliceToBobText"
        )
        val bobToAliceText = requireArgument(
            arguments.getString("bobToAliceText"),
            "bobToAliceText"
        )

        val sdk = WuKongEasySDK.getInstance()
        val receivedReply = CompletableDeferred<Message>()
        val listener = object : WuKongEventListener<Message> {
            override fun onEvent(data: Message) {
                val payload = data.payload as? Map<*, *> ?: return
                if (
                    data.fromUid == bobUid &&
                    payload["content"] == bobToAliceText &&
                    !receivedReply.isCompleted
                ) {
                    receivedReply.complete(data)
                }
            }
        }

        val config = WuKongConfig.Builder()
            .serverUrl("ws://10.0.2.2:5200")
            .uid(aliceUid)
            .token(aliceToken)
            .deviceFlag(WuKongDeviceFlag.APP)
            .debugLogging(false)
            .build()

        sdk.init(InstrumentationRegistry.getInstrumentation().targetContext, config)
        sdk.addEventListener(WuKongEvent.MESSAGE, listener)

        try {
            sdk.connect()
            assertTrue("Android SDK did not report connected", sdk.isConnected())

            val acknowledgment = sdk.send(
                channelId = bobUid,
                channelType = WuKongChannelType.PERSON,
                payload = mapOf("type" to 1, "content" to aliceToBobText)
            )
            assertTrue("SENDACK did not contain a sequence", acknowledgment.messageSeq > 0)

            val reply = withTimeout(30_000) { receivedReply.await() }
            assertTrue("Reply did not contain a sequence", reply.messageSeq > 0)
        } finally {
            sdk.removeEventListener(WuKongEvent.MESSAGE, listener)
            sdk.disconnect()
        }

        assertFalse("Android SDK remained connected after disconnect", sdk.isConnected())
        println(
            "ANDROID_RELEASE_SMOKE_PASS " +
                "package=com.githubim:easysdk-android:1.0.5 " +
                "alice-to-bob=true bob-to-alice=true disconnected=true"
        )
    }

    private fun requireArgument(value: String?, name: String): String {
        return value?.takeIf { it.isNotBlank() }
            ?: error("Missing instrumentation argument: $name")
    }
}
